src/util/src/WaveformUtil.cc

```diff
@@ -0,0 +1,251 @@
+    voltageWaveform.push_back((adcVal - pedestal) * voltageRes);
+  }
+
+  return voltageWaveform;
+}
+
+double CalculatePedestal(const std::vector<UShort_t>& waveform, int pedWindowLow, int pedWindowHigh) {
+  /*
+  Calculate the baseline (in ADC counts) in the window between low - high samples.
+  */
+  double pedestal = 0;
+
+  if (pedWindowLow > waveform.size()) {
+    Log::Die("WaveformUtil: Start of pedestal window must be before end of waveform.");
+  } else if (pedWindowLow >= pedWindowHigh) {
+    Log::Die("WaveformUtil: Start of pedestal window must be before end of pedestal window.");
+  } else if (pedWindowHigh > waveform.size()) {
+    Log::Die("WaveformUtil: End of pedestal window must be at most end of waveform.");
+  }
+
+  // Ensure end of pedestal window is less than waveform size
+  pedWindowHigh = (pedWindowHigh > waveform.size()) ? waveform.size() : pedWindowHigh;
+
+  for (int i = pedWindowLow; i < pedWindowHigh; i++) {
+    pedestal += waveform[i];
+  }
+  pedestal /= (pedWindowHigh - pedWindowLow);
+  return pedestal;
+}
+
+double CalculatePedestal(const std::vector<double>& waveform, int pedWindowLow, int pedWindowHigh) {
+  /*
+  Calculate the baseline (in mV) in the window between low - high samples.
+  */
+  double pedestal = 0;
+
+  if (pedWindowLow > waveform.size()) {
+    Log::Die("WaveformUtil: Start of pedestal window must be before end of waveform.");
+  } else if (pedWindowLow >= pedWindowHigh) {
+    Log::Die("WaveformUtil: Start of pedestal window must be before end of pedestal window.");
+  } else if (pedWindowHigh > waveform.size()) {
+    Log::Die("WaveformUtil: End of pedestal window must be at most end of waveform.");
+  }
+
+  // Ensure end of pedestal window is less than waveform size
+  pedWindowHigh = (pedWindowHigh > waveform.size()) ? waveform.size() : pedWindowHigh;
+
+  for (int i = pedWindowLow; i < pedWindowHigh; i++) {
+    pedestal += waveform[i];
+  }
+  pedestal /= (pedWindowHigh - pedWindowLow);
+  return pedestal;
+}
+
+std::pair<int, double> FindHighestPeak(const std::vector<double>& voltageWaveform) {
+  /*
+  Calculate the highest peak (in mV) and the corresponding sample.
+  */
+  double voltagePeak = INVALID;
+  int samplePeak = -INVALID;
+  for (int i = 0; i < voltageWaveform.size(); i++) {
+    double voltage = voltageWaveform[i];
+    // Downward going pulse
+    if (voltage < voltagePeak) {
+      voltagePeak = voltage;
+      samplePeak = i;
+    }
+  }
+  return std::make_pair(samplePeak, voltagePeak);
+}
+
+int GetThresholdCrossingBeforePeak(const std::vector<double>& waveform, int peakSample, double voltageThreshold,
+                                   int lookBack, double timeStep) {
+  /*
+  Identifies the sample at which threshold crossing occurs before a given peak
+   */
+  int thresholdCrossing = 0;
+  // Make sure we don't scan passed the beginning of the waveform
+  int lb = peakSample - int(lookBack / timeStep);
+  int back_window = (lb > 0) ? lb : 0;
+
```

Comment:
Let's also die if waveform.at(peakSample) is > threshold. I think this will guarantee that the time+1 error won't be emitted in the CFD, so we are at least internally seg-fault-proof.